### PR TITLE
docs: SQLAlchemyInitPlugin to SQLAlchemyPlugin fixes SerializationException for docs

### LINKED ISSUE #3464: Bug: SerializationException when running modeling-and-features demo from docs
### Description

Hi,

First of all thanks for developing Litestar, it proves to be a very useful piece of software here. Unfortunately I ran into an issue. 

I ran into an `msgspec_error` when requesting a page backed by sqlalchemy models which are connected via relationships. It seems that the database is correctly queried, a list of objects are returned, but then an exception is thrown when converting the objects to JSON.

I ran into this issue on my production code but when isolating an MCVE I noticed that the provided example in the documentation also shows the same unexpected behaviour on tested on two different machines.  One crucial change to the code is however adding an author to the database.

Since this is quite a show-stopper for me: Thanks in advance for having a look at this!

### URL to code causing the issue

https://docs.litestar.dev/2/tutorials/repository-tutorial/01-modeling-and-features.html

### MCVE

```python
from datetime import date
from typing import TYPE_CHECKING
from uuid import UUID

from sqlalchemy import ForeignKey, select
from sqlalchemy.orm import Mapped, mapped_column, relationship

from litestar import Litestar, get
from litestar.contrib.sqlalchemy.base import UUIDAuditBase, UUIDBase
from litestar.contrib.sqlalchemy.plugins import AsyncSessionConfig, SQLAlchemyAsyncConfig, SQLAlchemyInitPlugin

if TYPE_CHECKING:
    from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession


# the SQLAlchemy base includes a declarative model for you to use in your models.
# The `Base` class includes a `UUID` based primary key (`id`)
class Author(UUIDBase):
    name: Mapped[str]
    dob: Mapped[date]
    books: Mapped[list["Book"]] = relationship(back_populates="author", lazy="selectin")


# The `AuditBase` class includes the same UUID` based primary key (`id`) and 2
# additional columns: `created_at` and `updated_at`. `created_at` is a timestamp of when the
# record created, and `updated_at` is the last time the record was modified.
class Book(UUIDAuditBase):
    title: Mapped[str]
    author_id: Mapped[UUID] = mapped_column(ForeignKey("author.id"))
    author: Mapped[Author] = relationship(lazy="joined", innerjoin=True, viewonly=True)


session_config = AsyncSessionConfig(expire_on_commit=False)
sqlalchemy_config = SQLAlchemyAsyncConfig(
    connection_string="sqlite+aiosqlite:///test.sqlite", session_config=session_config
)  # Create 'async_session' dependency.
sqlalchemy_plugin = SQLAlchemyInitPlugin(config=sqlalchemy_config)


async def on_startup() -> None:
    """Initializes the database."""
    async with sqlalchemy_config.get_engine().begin() as conn:
        await conn.run_sync(UUIDBase.metadata.create_all)
        
    #crucially there needs to be an author in the table for the error to appear
    await conn.execute(Author.__table__.insert().values(name="F. Scott Fitzgerald"))


@get(path="/authors")
async def get_authors(db_session: "AsyncSession", db_engine: "AsyncEngine") -> list[Author]:
    """Interact with SQLAlchemy engine and session."""
    return list(await db_session.scalars(select(Author)))


app = Litestar(
    route_handlers=[get_authors],
    on_startup=[on_startup],
    plugins=[SQLAlchemyInitPlugin(config=sqlalchemy_config)],
    debug=True
)
```

### Steps to reproduce

```bash
1. Go to the https://docs.litestar.dev/2/tutorials/repository-tutorial/01-modeling-and-features.html page
2. Download the code
3. Run the demo with minimal requirements installed and go to http://localhost:8000/authors
4. See the error
```

### Screenshots

_No response_

### Logs

```bash
File "/usr/local/lib/python3.12/site-packages/litestar/serialization/msgspec_hooks.py", line 143, in encode_json
    raise SerializationException(str(msgspec_error)) from msgspec_error
litestar.exceptions.base_exceptions.SerializationException: Unsupported type: <class '__main__.Author'>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/litestar/serialization/msgspec_hooks.py", line 141, in encode_json
    return msgspec.json.encode(value, enc_hook=serializer) if serializer else _msgspec_json_encoder.encode(value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/litestar/serialization/msgspec_hooks.py", line 88, in default_serializer
    raise TypeError(f"Unsupported type: {type(value)!r}")
TypeError: Unsupported type: <class '__main__.Author'>
 
The above exception was the direct cause of the following exception:
 
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/litestar/middleware/exceptions/middleware.py", line 219, in __call__
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.12/site-packages/litestar/routes/http.py", line 82, in handle
    response = await self._get_response_for_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/litestar/routes/http.py", line 134, in _get_response_for_request
    return await self._call_handler_function(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/litestar/routes/http.py", line 158, in _call_handler_function
    response: ASGIApp = await route_handler.to_response(app=scope["app"], data=response_data, request=request)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/litestar/handlers/http_handlers/base.py", line 557, in to_response
    return await response_handler(app=app, data=data, request=request)  # type: ignore[call-arg]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/litestar/handlers/http_handlers/_utils.py", line 79, in handler
    return response.to_asgi_response(app=None, request=request, headers=normalize_headers(headers), cookies=cookies)  # pyright: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/litestar/response/base.py", line 451, in to_asgi_response
    body=self.render(self.content, media_type, get_serializer(type_encoders)),
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/litestar/response/base.py", line 392, in render
    return encode_json(content, enc_hook)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/litestar/serialization/msgspec_hooks.py", line 143, in encode_json
    raise SerializationException(str(msgspec_error)) from msgspec_error
litestar.exceptions.base_exceptions.SerializationException: Unsupported type: <class '__main__.Author'>
INFO:     127.0.0.1:44906 - "GET /authors HTTP/1.1" 500 Internal Server Error
```

### Litestar Version

2.8.2

### Platform

- [X] Linux
- [X] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/3464">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/3464/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by JorenSix**
## Description

This fixes a SerializationException when running the demo with dummy data included. The example now is fixed by using the SQLAlchemyPlugin in stead of the SQLAlchemyInitPlugin and dummy data is added which immediately shows the expected behaviour: a JSON list with authors.

See issue #3464

-

## Closes

Closes #3464

### CODE REVIEW

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 1]
```diff
@@ -34,13 +35,21 @@ class Book(UUIDAuditBase):
...
 
 async def on_startup() -> None:
-    """Initializes the database."""
+    """Initializes the database and add some dummy data."""
     async with sqlalchemy_config.get_engine().begin() as conn:
         await conn.run_sync(UUIDBase.metadata.create_all)
+    async with sqlalchemy_config.get_session() as session:
+        statement = select(func.count()).select_from(Author)
+        count = await session.execute(statement)
+        if 0 == count.scalar():
```
Comment (peterschutt) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:48]:
```suggestion
        if not count.scalar():
```

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 2]
```diff
@@ -34,13 +35,21 @@ class Book(UUIDAuditBase):
 sqlalchemy_config = SQLAlchemyAsyncConfig(
     connection_string="sqlite+aiosqlite:///test.sqlite", session_config=session_config
 )  # Create 'async_session' dependency.
-sqlalchemy_plugin = SQLAlchemyInitPlugin(config=sqlalchemy_config)
+sqlalchemy_plugin = SQLAlchemyPlugin(config=sqlalchemy_config)
```
Comment (peterschutt) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:38]:
```suggestion
```

This plugin is instantiated when we create the app below.

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 3]
```diff
@@ -34,13 +35,21 @@ class Book(UUIDAuditBase):
...
 
 
 async def on_startup() -> None:
-    """Initializes the database."""
+    """Initializes the database and add some dummy data."""
```
Comment (peterschutt) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:42]:
```suggestion
    """Initializes the database and adds some dummy data."""
```

**Review by peterschutt — CHANGES_REQUESTED** [review]
Thanks @JorenSix!

Beyond the missing serialization plugin, I like the idea of seeding the database with some test data!

We really should have a test for the example in `tests/examples/test_contrib/test_sqlalchemy`, would you mind adding one as part of this PR? It should hit the `GET /authors` endpoint and test the json response.

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 4]
```diff
@@ -34,13 +35,21 @@ class Book(UUIDAuditBase):
 sqlalchemy_config = SQLAlchemyAsyncConfig(
     connection_string="sqlite+aiosqlite:///test.sqlite", session_config=session_config
```
Comment (cofin) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:36]:
```suggestion
    connection_string="sqlite+aiosqlite:///test.sqlite", session_config=session_config, create_all=True
```

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 5]
```diff
@@ -34,13 +35,21 @@ class Book(UUIDAuditBase):
...
 )  # Create 'async_session' dependency.
-sqlalchemy_plugin = SQLAlchemyInitPlugin(config=sqlalchemy_config)
+sqlalchemy_plugin = SQLAlchemyPlugin(config=sqlalchemy_config)
 
 
 async def on_startup() -> None:
-    """Initializes the database."""
+    """Initializes the database and add some dummy data."""
     async with sqlalchemy_config.get_engine().begin() as conn:
         await conn.run_sync(UUIDBase.metadata.create_all)
```
Comment (cofin) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:44]: This can be removed if we change change the `create_all=True` in the `SQLAlchemyConfig`
```suggestion
```

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py (hunk 4 above)
Comment (cofin) [reply docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:36]: This flag will now work since we are swapping to the main `SQLAlchemyPlugin` here.

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 6]
```diff
@@ -52,5 +61,5 @@ async def get_authors(db_session: "AsyncSession", db_engine: "AsyncEngine") -> l
 app = Litestar(
     route_handlers=[get_authors],
     on_startup=[on_startup],
```
Comment (cofin) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:63]:
```suggestion
```

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 7]
```diff
@@ -34,13 +35,21 @@ class Book(UUIDAuditBase):
 sqlalchemy_config = SQLAlchemyAsyncConfig(
     connection_string="sqlite+aiosqlite:///test.sqlite", session_config=session_config
 )  # Create 'async_session' dependency.
-sqlalchemy_plugin = SQLAlchemyInitPlugin(config=sqlalchemy_config)
+sqlalchemy_plugin = SQLAlchemyPlugin(config=sqlalchemy_config)
 
 
 async def on_startup() -> None:
```
Comment (cofin) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:41]: This whole function can be removed if we change change the `create_all=True` in the `SQLAlchemyConfig`

Comment (Alc-Alc) [reply docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:40]: @cofin I think your intention was to just remove the first two lines (the ones that create the tables) and not the entire function because @JorenSix is adding some data here and removing this hook would prevent that.

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 8]
```diff
@@ -32,15 +33,22 @@ class Book(UUIDAuditBase):
...
+    connection_string="sqlite+aiosqlite:///test.sqlite", session_config=session_config, create_all=True
 )  # Create 'async_session' dependency.
-sqlalchemy_plugin = SQLAlchemyInitPlugin(config=sqlalchemy_config)
 
 
 async def on_startup() -> None:
-    """Initializes the database."""
+    """Initializes the database and adds some dummy data."""
     async with sqlalchemy_config.get_engine().begin() as conn:
         await conn.run_sync(UUIDBase.metadata.create_all)
```
Comment (Alc-Alc) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:43]:
```suggestion
```
I think this is what he meant

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 9]
```diff
@@ -51,6 +59,5 @@ async def get_authors(db_session: "AsyncSession", db_engine: "AsyncEngine") -> l
 
 app = Litestar(
     route_handlers=[get_authors],
-    on_startup=[on_startup],
```
Comment (Alc-Alc) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:54]: this should not be removed if the data is to be seeded (which this PR adds)

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py (hunk 8 above)
Comment (JorenSix) [reply docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py]: See the commit below, I have indeed solved it by creating the metadata using the parameter, the data via the on_startup.

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py (hunk 7 above)
Comment (JorenSix) [reply docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:40]: See below, indeed now I create the metadata via the parameter and the data via the on_startup. Thanks for the suggestion!

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py (hunk 9 above)
Comment (JorenSix) [reply docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:54]: Indeed, see below!

**Comment by JorenSix** [comment]
I have modified the example using the suggestions above. The metadata is now created automatically, the data via on_startup.

I also created a new unit test file which checks if the response is 200 and a non empty list of authors is returned. This works with recent python versions.

Unfortunately this fails for older python versions (3.9, 3.10), which immediately shows why it is indeed a good idea to provide a unit test. The `litestar.exceptions.base_exceptions.SerializationException` looks similar to me as the original associated issue #3464 


````
Traceback (most recent call last):
  File "/workspaces/litestar/litestar/middleware/_internal/exceptions/middleware.py", line 158, in __call__
    await self.app(scope, receive, capture_response_started)
  File "/workspaces/litestar/litestar/_asgi/asgi_router.py", line 99, in __call__
    await asgi_app(scope, receive, send)
  File "/workspaces/litestar/litestar/routes/http.py", line 80, in handle
    response = await self._get_response_for_request(
  File "/workspaces/litestar/litestar/routes/http.py", line 132, in _get_response_for_request
    return await self._call_handler_function(
  File "/workspaces/litestar/litestar/routes/http.py", line 156, in _call_handler_function
    response: ASGIApp = await route_handler.to_response(app=scope["app"], data=response_data, request=request)
  File "/workspaces/litestar/litestar/handlers/http_handlers/base.py", line 557, in to_response
    return await response_handler(app=app, data=data, request=request)  # type: ignore[call-arg]
  File "/workspaces/litestar/litestar/handlers/http_handlers/_utils.py", line 79, in handler
    return response.to_asgi_response(app=None, request=request, headers=normalize_headers(headers), cookies=cookies)  # pyright: ignore
  File "/workspaces/litestar/litestar/response/base.py", line 451, in to_asgi_response
    body=self.render(self.content, media_type, get_serializer(type_encoders)),
  File "/workspaces/litestar/litestar/response/base.py", line 392, in render
    return encode_json(content, enc_hook)
  File "/workspaces/litestar/litestar/serialization/msgspec_hooks.py", line 152, in encode_json
    raise SerializationException(str(msgspec_error)) from msgspec_error
litestar.exceptions.base_exceptions.SerializationException: Unsupported type: <class 'docs.examples.contrib.sqlalchemy.sqlalchemy_declarative_models.Book'>
`````

Feedback or instructions on how to proceed are welcome.

As an aside: I already learned a lot about the litestar CI infra and the use of the `.devcontainer` folder. Very practical indeed to e.g. switch python versions and quickly run unit tests.

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 10]
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations # Required for forward class references, see https://peps.python.org/pep-0484/#forward-references
```
Comment (Alc-Alc) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py]: IMO, I don't think the comment explaining is needed, it would raise some questions if this is the only place we define them and not for other examples

File: docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py [hunk 11]
```diff
@@ -18,7 +21,7 @@
 class Author(UUIDBase):
     name: Mapped[str]
     dob: Mapped[date]
-    books: Mapped[list["Book"]] = relationship(back_populates="author", lazy="selectin")
+    books: Mapped[list[Book]] = relationship(back_populates="author", lazy="selectin")
```
Comment (peterschutt) [inline docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:24]: To run on 3.8 this would have to be:

```py
from typing import List
...
class Author(UUIDBase):
    name: Mapped[str]
    dob: Mapped[date]
    books: Mapped[List[Book]] = relationship(back_populates="author", lazy="selectin")
```

Comment (JorenSix) [reply docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:24]: I seem to be running in the following ruff error:

`docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:24:19: UP006 Use `list` instead of `List` for type annotation`

I am not sure why other scripts, which also use `List` do not trigger the same ruff errors.  See e.g. `/docs/examples/data_transfer_objects/factory/included_fields.py` 

Any tips?

Comment (peterschutt) [reply docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:24]: Yeah, we disable that lint for the file in pyproject:
https://github.com/litestar-org/litestar/blob/d75efa7e79be5a252920aa29f204eb0b310d62a4/pyproject.toml#L397

Comment (JorenSix) [reply docs/examples/contrib/sqlalchemy/sqlalchemy_declarative_models.py:24]: Ok thanks for the info

**Comment by JorenSix** [comment]
Thanks for following up and the quick feedback. The process and the infrastructure of litestar seems exemplary. I learned a lot about how you handle things here. Thanks again.

**Comment by peterschutt** [comment]
@all-contributors add @JorenSix for docs
